pkg/k8s/watchers/node.go

```diff
@@ -117,3 +135,109 @@ func (k *K8sWatcher) GetK8sNode(_ context.Context, nodeName string) (*v1.Node, e
 	}
 	return nodeInterface.(*v1.Node).DeepCopy(), nil
 }
+
+// CiliumNodeUpdater implements the subscriber.Node interface and is used
+// to keep CiliumNode objects in sync with the node ones.
+type CiliumNodeUpdater struct {
+	k8sWatcher    *K8sWatcher
+	nodeDiscovery *nodediscovery.NodeDiscovery
+}
+
+func NewCiliumNodeUpdater(k8sWatcher *K8sWatcher, nodeDiscovery *nodediscovery.NodeDiscovery) *CiliumNodeUpdater {
+	return &CiliumNodeUpdater{
+		k8sWatcher:    k8sWatcher,
+		nodeDiscovery: nodeDiscovery,
+	}
+}
+
+func (u *CiliumNodeUpdater) OnAddNode(newNode *v1.Node, swg *lock.StoppableWaitGroup) error {
+	u.updateCiliumNode(u.nodeDiscovery, newNode)
+
+	return nil
+}
+
+func (u *CiliumNodeUpdater) OnUpdateNode(oldNode, newNode *v1.Node, swg *lock.StoppableWaitGroup) error {
+	u.updateCiliumNode(u.nodeDiscovery, newNode)
+
+	return nil
+}
+
+func (u *CiliumNodeUpdater) OnDeleteNode(*v1.Node, *lock.StoppableWaitGroup) error {
+	return nil
+}
+
+func (u *CiliumNodeUpdater) updateCiliumNode(nodeDiscovery *nodediscovery.NodeDiscovery, node *v1.Node) {
+	controllerName := fmt.Sprintf("sync-node-with-ciliumnode (%v)", node.Name)
+
+	nodeSlim := k8s.ConvertToNode(node.DeepCopy()).(*slim_corev1.Node)
+	k8sNodeParsed := k8s.ParseNode(nodeSlim, source.Unspec)
+
+	k8sCM.UpdateController(controllerName,
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) (err error) {
+				if option.Config.KVStore != "" {
+					if nodeDiscovery.Registrar.SharedStore == nil {
+						return fmt.Errorf("node registrar is not yet initialized")
+					}
+
+					if err := nodeDiscovery.Registrar.UpdateLocalKeySync(k8sNodeParsed); err != nil {
+						return fmt.Errorf("failed to update KV store entry: %s", err)
+					}
+				} else {
+					u.k8sWatcher.ciliumNodeStoreMU.Lock()
+					if u.k8sWatcher.ciliumNodeStore == nil {
+						u.k8sWatcher.ciliumNodeStoreMU.Unlock()
+						return errors.New("CiliumNode cache store not yet initialized")
+					}
+					u.k8sWatcher.ciliumNodeStoreMU.Unlock()
+
+					ciliumNodeInterface, exists, err := u.k8sWatcher.ciliumNodeStore.GetByKey(node.Name)
+					if err != nil {
+						return fmt.Errorf("failed to get CiliumNode resource from cache store: %w", err)
+					}
+					if !exists {
+						return nil
+					}
+
+					ciliumNode := ciliumNodeInterface.(*ciliumv2.CiliumNode).DeepCopy()
+
+					ciliumNode.Labels = node.GetLabels()
+
+					k8sNodeAddresses := k8sNodeParsed.IPAddresses
+
+					ciliumNode.Spec.Addresses = []ciliumv2.NodeAddress{}
+					for _, k8sAddress := range k8sNodeAddresses {
+						k8sAddressStr := k8sAddress.IP.String()
+						ciliumNode.Spec.Addresses = append(ciliumNode.Spec.Addresses, ciliumv2.NodeAddress{
+							Type: k8sAddress.Type,
+							IP:   k8sAddressStr,
+						})
+					}
+
+					for _, address := range nodeDiscovery.LocalNode.IPAddresses {
```

Comment:
[1] 

I think we need to be careful here because `LocalNode` can be written to async from somewhere else. Especially with @gandro's [PR](https://github.com/cilium/cilium/pull/18887) where this variable will change at runtime, we'll need a way to retrieve state from it safely. Maybe it's worth defining an interface to abstract away the underlying object?